// rolldown/dist/types/cli/arguments/alias.js
export const alias = {
    config: {
        abbreviation: 'c',
        hint: 'filename',
    },
    help: {
        abbreviation: 'h',
    },
    version: {
        abbreviation: 'v',
    },
    watch: {
        abbreviation: 'w',
    },
    dir: {
        abbreviation: 'd',
    },
    file: {
        abbreviation: 'o',
    },
    external: {
        abbreviation: 'e',
    },
    format: {
        abbreviation: 'f',
    },
    name: {
        abbreviation: 'n',
    },
    globals: {
        abbreviation: 'g',
    },
    sourcemap: {
        abbreviation: 's',
        default: true,
    },
    minify: {
        abbreviation: 'm',
    },
    platform: {
        abbreviation: 'p',
    },
    assetFileNames: {
        hint: 'name',
    },
    chunkFileNames: {
        hint: 'name',
    },
    entryFileNames: {
        hint: 'name',
    },
    externalLiveBindings: {
        default: true,
        reverse: true,
    },
    treeshake: {
        default: true,
        reverse: true,
    },
    moduleTypes: {
        hint: 'types',
    },
};
